test/solver_tests/percolation_test.py

```diff
@@ -0,0 +1,200 @@
+"""
+Test PercolationSolver in Cassiopeia.solver.
+"""
+
+import unittest
+
+import networkx as nx
+import numpy as np
+import pandas as pd
+from typing import Dict, List, Optional
+
+import cassiopeia as cas
+from cassiopeia.data import utilities as data_utilities
+from cassiopeia.solver.VanillaGreedySolver import VanillaGreedySolver
+from cassiopeia.solver.NeighborJoiningSolver import NeighborJoiningSolver
+from cassiopeia.solver.PercolationSolver import PercolationSolver
+from cassiopeia.solver import dissimilarity_functions
+
+
+def neg_hamming_similarity_without_missing(
+    s1: np.array,
+    s2: np.array,
+    missing_state_indicator: int,
+    weights: Optional[Dict[int, Dict[int, float]]] = None,
+) -> float:
+    """A function to return the number of (non-missing) character/state
+    mutations shared by two samples.
+
+    Args:
+        s1: Character states of the first sample
+        s2: Character states of the second sample
+        missing_state_indicator: The character representing missing values
+        weights: A set of optional weights to weight the similarity of a mutation
+    Returns:
+        The number of shared mutations between two samples, weighted or unweighted
+    """
+
+    # TODO Optimize this using masks
+    similarity = 0
+    for i in range(len(s1)):
+
+        if (
+            s1[i] == missing_state_indicator
+            or s2[i] == missing_state_indicator
+            or s1[i] == 0
+            or s2[i] == 0
+        ):
+            continue
+
+        if s1[i] == s2[i]:
+            if weights:
+                similarity += weights[i][s1[i]]
+            else:
+                similarity += 1
+
+    return -1 * similarity
+
+
+class PercolationSolverTest(unittest.TestCase):
+    def test_NJ_negative_similarity(self):
+
+        cm = pd.DataFrame(
+            [
+                [5, 3, 0, 0, 0],
+                [0, 3, 4, 2, 1],
+                [5, 0, 0, 0, 1],
+                [5, 0, 4, 2, 0],
+                [5, 0, 0, 0, 0],
+                [5, 0, 0, 0, 0],
+            ]
+        )
+        p_tree = cas.data.CassiopeiaTree(
+            cm, missing_state_indicator=-1
+        )
+
+        joining_solver = NeighborJoiningSolver(dissimilarity_function = neg_hamming_similarity_without_missing, add_root = True)
+        psolver = PercolationSolver(joining_solver = joining_solver)
+        psolver.solve(p_tree)
+        expected_newick_string = "((1,3),(2,0,(4,5)));"
```

Comment:
I've actually found that perhaps a better way to test for structure is to create a new networkx graph with the correct structure and make sure that the triplets correct or RF or LCA structure is perfectly in agreement. The newick strings can be arbitrarily formed (for example - if you switched `(1,3)` to be `(3,1)` the structure would be the same).